## Analyzing US Lottery Data
#### Mega Millions draws from 2002 to 2025

In [1]:
import pandas as pd
from statistics import mode

In [2]:
# Load the dataset
file_path = './Lottery_Mega_Millions_Winning_Numbers__Beginning_2002.csv'
data = pd.read_csv(file_path)

# Look at the first few rows of the dataset
print(data.head())

# Show information about the dataset
print(data.info())

    Draw Date Winning Numbers  Mega Ball  Multiplier
0  09/25/2020  20 36 37 48 67         16         2.0
1  09/29/2020  14 39 43 44 67         19         3.0
2  10/02/2020  09 38 47 49 68         25         2.0
3  10/06/2020  15 16 18 39 59         17         3.0
4  10/09/2020  05 11 25 27 64         13         2.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2396 entries, 0 to 2395
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Draw Date        2396 non-null   object 
 1   Winning Numbers  2396 non-null   object 
 2   Mega Ball        2396 non-null   int64  
 3   Multiplier       1483 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 75.0+ KB
None


In [3]:
# Rename columns to lowercase and replace spaces with underscores
data.columns = [name for name in data.columns.str.replace(' ', '_').str.lower()]

In [4]:
# Show unique values in the 'multiplier' column
display(data['multiplier'].unique())

# Fill missing values in the 'multiplier' column with 1 as there is no multiplier
data['multiplier'] = data['multiplier'].fillna(1)
# Convert the 'multiplier' column to integer type
data['multiplier'] = data['multiplier'].astype(int)

# Show unique values in the 'multiplier' column
display(data['multiplier'].unique())

array([ 2.,  3.,  5.,  4., nan])

array([2, 3, 5, 4, 1])

In [5]:
# Convert the 'draw_date' column to datetime format
data['draw_date'] = pd.to_datetime(data['draw_date'], format='%m/%d/%Y')

In [6]:
# Convertng the 'winning_numbers' column to a list of integers
def convert_to_list_of_ints(s):
    return [int(num) for num in s.split()]
data['winning_numbers'] = data['winning_numbers'].apply(convert_to_list_of_ints)

# Transform the 'winning_numbers' column into separate columns
data[['wn_1', 'wn_2', 'wn_3', 'wn_4', 'wn_5']] = data['winning_numbers'].apply(pd.Series)

# Delete the original 'winning_numbers' column
data = data.drop(columns=['winning_numbers'])


In [7]:
# Showing informations about the dataset
display(data.head())
display(data.info())
display(data.describe())

draw_date  mega_ball  multiplier  wn_1  wn_2  wn_3  wn_4  wn_5
0 2020-09-25         16           2    20    36    37    48    67
1 2020-09-29         19           3    14    39    43    44    67
2 2020-10-02         25           2     9    38    47    49    68
3 2020-10-06         17           3    15    16    18    39    59
4 2020-10-09         13           2     5    11    25    27    64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2396 entries, 0 to 2395
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   draw_date   2396 non-null   datetime64[ns]
 1   mega_ball   2396 non-null   int64         
 2   multiplier  2396 non-null   int32         
 3   wn_1        2396 non-null   int64         
 4   wn_2        2396 non-null   int64         
 5   wn_3        2396 non-null   int64         
 6   wn_4        2396 non-null   int64         
 7   wn_5        2396 non-null   int64         
dtypes: datetime64[ns](1), int32(1), int64(6)
memory usage: 140.5 KB


None

draw_date    mega_ball   multiplier         wn_1  \
count                           2396  2396.000000  2396.000000  2396.000000   
mean   2013-11-13 05:58:11.819699456    17.826795     2.416945    10.424040   
min              2002-05-17 00:00:00     1.000000     1.000000     1.000000   
25%              2008-02-14 06:00:00     8.000000     1.000000     4.000000   
50%              2013-11-17 00:00:00    15.000000     2.000000     8.000000   
75%              2019-08-13 18:00:00    25.000000     4.000000    15.000000   
max              2025-05-09 00:00:00    52.000000     5.000000    56.000000   
std                              NaN    12.643153     1.357646     8.383481   

              wn_2         wn_3         wn_4         wn_5  
count  2396.000000  2396.000000  2396.000000  2396.000000  
mean     21.048414    31.706594    42.421119    53.235392  
min       2.000000     3.000000     5.000000    14.000000  
25%      13.000000    23.000000    34.000000    46.000000  
50%      19.500000    31.000000    42.000000    53.000000  
75%      28.000000    40.000000    51.000000    62.250000  
max      66.000000    68.000000    74.000000    75.000000  
std      11.054815    12.301751    12.309069    11.395513

In [8]:
# Showing the most common winning numbers in order of draw
print(mode(data['wn_1']), mode(data['wn_2']), mode(data['wn_3']), mode(data['wn_4']), mode(data['wn_5']), mode(data['mega_ball']))

2 17 31 46 52 9


In [9]:
# Showing the first 5 most common winning numbers for eash draw
display(data['wn_1'].value_counts().index[:5].to_list(), 
        data['wn_2'].value_counts().index[:5].to_list(), 
        data['wn_3'].value_counts().index[:5].to_list(), 
        data['wn_4'].value_counts().index[:5].to_list(), 
        data['wn_5'].value_counts().index[:5].to_list(), 
        data['mega_ball'].value_counts().index[:5].to_list())

[2, 3, 1, 4, 7]

[17, 15, 10, 13, 20]

[31, 26, 27, 24, 29]

[46, 42, 38, 48, 40]

[52, 51, 56, 50, 55]

[9, 7, 3, 10, 13]

In [10]:
# Show how % of the winning numbers are in the first 5 most common winning numbers
print(data['wn_1'].value_counts()[:5].sum() / data['wn_1'].count() * 100)
print(data['wn_2'].value_counts()[:5].sum() / data['wn_2'].count() * 100)
print(data['wn_3'].value_counts()[:5].sum() / data['wn_3'].count() * 100)
print(data['wn_4'].value_counts()[:5].sum() / data['wn_4'].count() * 100)
print(data['wn_5'].value_counts()[:5].sum() / data['wn_5'].count() * 100)

35.14190317195325
18.948247078464107
16.569282136894824
17.487479131886477
22.20367278797997


In [11]:
# Show the top 5 most common winning numbers for all draws
all_numbers = pd.concat([data['wn_1'], data['wn_2'], data['wn_3'], data['wn_4'], data['wn_5']])

# Show how % of the winning numbers are in the first 5 most common winning numbers
display(all_numbers.value_counts().index[:5].to_list())
all_numbers.value_counts()[:5].sum() / all_numbers.count() * 100


[31, 10, 20, 14, 17]

9.332220367278799

It was observed that 5 numbers hold about 9.33% of the total numbers drawn since, we also got a 5x5 set with which has about 20% combined draws, where the lines are the order of the draw and the columns are the numbers drawn, we have the top 1 numbers by order of draw that have never been drawn together before, which can be a big bet if we were not talking about a random draw.